examples/pytorch/ogb_lsc/MAG240M/experimental/train_unified_tensor.py

```diff
@@ -0,0 +1,233 @@
+            self.convs.append(nn.ModuleList([
+                dglnn.GATConv(hidden_channels, hidden_channels // num_heads, num_heads, allow_zero_in_degree=True)
+                for _ in range(num_etypes)
+            ]))
+            self.norms.append(nn.BatchNorm1d(hidden_channels))
+            self.skips.append(nn.Linear(hidden_channels, hidden_channels))
+            
+        self.mlp = nn.Sequential(
+            nn.Linear(hidden_channels, hidden_channels),
+            nn.BatchNorm1d(hidden_channels),
+            nn.ReLU(),
+            nn.Dropout(dropout),
+            nn.Linear(hidden_channels, out_channels)
+        )
+        self.dropout = nn.Dropout(dropout)
+        
+        self.hidden_channels = hidden_channels
+        self.pred_ntype = pred_ntype
+        self.num_etypes = num_etypes
+        
+    def forward(self, mfgs, x):
+        for i in range(len(mfgs)):
+            mfg = mfgs[i]
+            x_dst = x[:mfg.num_dst_nodes()]
+            n_src = mfg.num_src_nodes()
+            n_dst = mfg.num_dst_nodes()
+            mfg = dgl.block_to_graph(mfg)
+            x_skip = self.skips[i](x_dst)
+            for j in range(self.num_etypes):
+                subg = mfg.edge_subgraph(mfg.edata['etype'] == j, relabel_nodes=False)
+                x_skip += self.convs[i][j](subg, (x, x_dst)).view(-1, self.hidden_channels)
+            x = self.norms[i](x_skip)
+            x = F.elu(x)
+            x = self.dropout(x)
+        return self.mlp(x)
+
+
+class ExternalNodeCollator(dgl.dataloading.NodeCollator):
+    def __init__(self, g, idx, sampler, offset, label):
+        super().__init__(g, idx, sampler)
+        self.offset = offset
+        self.label = label
+
+    def collate(self, items):
+        input_nodes, output_nodes, mfgs = super().collate(items)
+        # Copy input features
+        mfgs[-1].dstdata['y'] = torch.LongTensor(self.label[output_nodes - self.offset])
+        return input_nodes, output_nodes, mfgs
+
+def train(args, dataset, g, feats, paper_offset):
+    print('Loading masks and labels')
+    train_idx = torch.LongTensor(dataset.get_idx_split('train')) + paper_offset
+    valid_idx = torch.LongTensor(dataset.get_idx_split('valid')) + paper_offset
+    label = dataset.paper_label
+
+    feats = dgl.contrib.UnifiedTensor(feats, device=torch.device("cuda:0"))
+
+    print('Initializing dataloader...')
+    sampler = dgl.dataloading.MultiLayerNeighborSampler([15, 25])
+    train_collator = ExternalNodeCollator(g, train_idx, sampler, paper_offset, label)
+    valid_collator = ExternalNodeCollator(g, valid_idx, sampler, paper_offset, label)
+    train_dataloader = torch.utils.data.DataLoader(
+        train_collator.dataset,
+        batch_size=1024,
+        shuffle=True,
+        drop_last=False,
+        collate_fn=train_collator.collate,
+        num_workers=4
+    )
+    valid_dataloader = torch.utils.data.DataLoader(
+        valid_collator.dataset,
+        batch_size=1024,
+        shuffle=True,
```

Comment:
```suggestion
        shuffle=False,
```
We do not need to shuffle the data for validation.

Reply:
This should be also addressed by migrating to GraphSAGE example.